File: src/lib/scoring.ts
import type { Doc } from "@/convex/_generated/dataModel";

export interface ScoreInput {
  ppeChecks: Doc<"ppeChecks">[];
  alerts: Doc<"alerts">[];
  session: {
    completedSteps: number[];
    skippedSteps: number[];
    incorrectSteps: number[];
  } | null;
  totalSteps: number;
  routeDeviations: number;
  gpsPoints: number;
}

export interface ScoreResult {
  ppeCompliance: number;
  sopCompliance: number;
  safetyCompliance: number;
  routeCompliance: number;
  sequenceCompliance: number;
  toolCompliance: number;
  overallScore: number;
}

/** Configurable weights (sums to 1). */
export const SCORE_WEIGHTS = {
  ppe: 0.2,
  sop: 0.3,
  safety: 0.2,
  route: 0.1,
  sequence: 0.15,
  tool: 0.05,
} as const;

const clamp = (n: number) => Math.max(0, Math.min(100, Math.round(n)));

export function computeJobScore(input: ScoreInput): ScoreResult {
  // PPE compliance: mean of per-check detected-item ratio
  let ppeCompliance = 100;
  if (input.ppeChecks.length > 0) {
    const ratios = input.ppeChecks.map((c) => {
      const items = Object.values(c.items);
      if (items.length === 0) return 0;
      return items.filter((i) => i.detected).length / items.length;
    });
    ppeCompliance = (ratios.reduce((a, b) => a + b, 0) / ratios.length) * 100;
  }

  // SOP compliance: completed steps minus penalties for skips/incorrect
  const totalSteps = Math.max(1, input.totalSteps);
  const completed = input.session?.completedSteps.length ?? 0;
  const skipped = input.session?.skippedSteps.length ?? 0;
  const incorrect = input.session?.incorrectSteps.length ?? 0;
  let sopCompliance = (completed / totalSteps) * 100 - skipped * 12 - incorrect * 8;

  // Safety compliance: unresolved high/critical alerts deduct
  let safetyCompliance = 100;
  for (const a of input.alerts) {
    if (a.resolved) continue;
    if (a.severity === "CRITICAL") safetyCompliance -= 25;
    else if (a.severity === "HIGH") safetyCompliance -= 15;
    else if (a.severity === "MEDIUM") safetyCompliance -= 8;
  }

  // Route compliance: share of GPS points off the planned route
  let routeCompliance = 100;
  if (input.gpsPoints > 0) {
    routeCompliance -= (input.routeDeviations / input.gpsPoints) * 100 * 2;
  }

  // Sequence compliance: wrong-order / skipped steps
  let sequenceCompliance = 100 - incorrect * 12 - skipped * 10;

  // Tool compliance: wrong-tool alerts
  const wrongTools = input.alerts.filter((a) => a.type === "WRONG_TOOL").length;
  let toolCompliance = 100 - wrongTools * 15;

  const overallScore = clamp(
    clamp(ppeCompliance) * SCORE_WEIGHTS.ppe +
      clamp(sopCompliance) * SCORE_WEIGHTS.sop +
      clamp(safetyCompliance) * SCORE_WEIGHTS.safety +
      clamp(routeCompliance) * SCORE_WEIGHTS.route +
      clamp(sequenceCompliance) * SCORE_WEIGHTS.sequence +
      clamp(toolCompliance) * SCORE_WEIGHTS.tool,
  );

  return {
    ppeCompliance: clamp(ppeCompliance),
    sopCompliance: clamp(sopCompliance),
    safetyCompliance: clamp(safetyCompliance),
    routeCompliance: clamp(routeCompliance),
    sequenceCompliance: clamp(sequenceCompliance),
    toolCompliance: clamp(toolCompliance),
    overallScore,
  };
}

File: ai-maintenance-backend/app/sop/scoring.py
"""Quality scoring engine.

Computes PPE, SOP, safety, route, sequence and tool compliance and rolls them
into a 0-100 overall score. Weights are configurable via settings.
"""
from __future__ import annotations

from typing import Dict, List, Optional

from app.config import get_settings


def _clamp(value: float) -> float:
    return max(0.0, min(100.0, round(value, 1)))


class ScoringEngine:
    def __init__(self) -> None:
        s = get_settings()
        self.weights = {
            "ppe": s.weight_ppe,
            "sop": s.weight_sop,
            "safety": s.weight_safety,
            "route": s.weight_route,
            "sequence": s.weight_sequence,
            "tool": s.weight_tool,
        }

    def compute(
        self,
        *,
        ppe_checks: Optional[List[Dict]] = None,
        sop_completed: int = 0,
        sop_total: int = 1,
        skipped_steps: int = 0,
        incorrect_steps: int = 0,
        alerts: Optional[List[Dict]] = None,
        route_deviations: int = 0,
        gps_points: int = 0,
        wrong_tools: int = 0,
    ) -> Dict[str, float]:
        """Return dict of compliance scores + overall_score (0-100)."""
        ppe_checks = ppe_checks or []
        alerts = alerts or []

        # PPE compliance: mean of per-check detected ratio
        if ppe_checks:
            ratios = []
            for check in ppe_checks:
                items = list(check.get("items", {}).values())
                if items:
                    ratios.append(sum(1 for i in items if i.get("detected")) / len(items))
            ppe = (sum(ratios) / len(ratios)) * 100 if ratios else 100.0
        else:
            ppe = 100.0

        # SOP compliance
        sop = (sop_completed / max(1, sop_total)) * 100 - skipped_steps * 12 - incorrect_steps * 8

        # Safety compliance
        safety = 100.0
        for alert in alerts:
            if alert.get("resolved"):
                continue
            severity = alert.get("severity")
            if severity == "CRITICAL":
                safety -= 25
            elif severity == "HIGH":
                safety -= 15
            elif severity == "MEDIUM":
                safety -= 8

        # Route compliance
        route = 100.0
        if gps_points > 0:
            route -= (route_deviations / gps_points) * 100 * 2

        # Sequence compliance
        sequence = 100.0 - incorrect_steps * 12 - skipped_steps * 10

        # Tool compliance
        tool = 100.0 - wrong_tools * 15

        parts = {
            "ppe_compliance": _clamp(ppe),
            "sop_compliance": _clamp(sop),
            "safety_compliance": _clamp(safety),
            "route_compliance": _clamp(route),
            "sequence_compliance": _clamp(sequence),
            "tool_compliance": _clamp(tool),
        }
        overall = sum(parts[k] * w for k, w in [
            ("ppe_compliance", self.weights["ppe"]),
            ("sop_compliance", self.weights["sop"]),
            ("safety_compliance", self.weights["safety"]),
            ("route_compliance", self.weights["route"]),
            ("sequence_compliance", self.weights["sequence"]),
            ("tool_compliance", self.weights["tool"]),
        ])
        parts["overall_score"] = _clamp(overall)
        return parts
